ssm_jax/hmm/core.py

```diff
@@ -0,0 +1,226 @@
+
+        # Renormalize
+        norm = backward_pred_probs.sum()
+        backward_pred_probs /= norm
+
+        # Update the log log_normalizer
+        log_normalizer += jnp.log(norm) + ll_max
+
+        return (log_normalizer, backward_pred_probs), backward_pred_probs
+
+    num_timesteps, num_states = log_likelihoods.shape
+    carry = (0.0, jnp.ones(num_states))
+    (log_normalizer, _), backward_pred_probs = lax.scan(
+        _step, carry, jnp.arange(num_timesteps)[::-1])
+    return log_normalizer, backward_pred_probs
+
+
+def hmm_two_filter_smoother(initial_distribution,
+                            transition_matrix,
+                            log_likelihoods):
+    # Run the filters forward and backward
+    ll, filtered_probs = hmm_filter(initial_distribution,
+                                    transition_matrix,
+                                    log_likelihoods)
+
+    _, backward_pred_probs = hmm_backward_filter(transition_matrix,
+                                                 log_likelihoods)
+
+    # Compute smoothed probabilities
+    smoothed_probs = filtered_probs * backward_pred_probs
+    norm = smoothed_probs.sum(axis=1)
+    smoothed_probs /= norm
+
+    return smoothed_probs, ll
+
+
+def hmm_smoother(initial_distribution,
+                 transition_matrix,
+                 log_likelihoods):
+    """_summary_
+
+    Args:
+        hmm_params (_type_): _description_
+        log_likelihoods (_type_): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    # Run the HMM filter
+    log_normalizer, filtered_probs = hmm_filter(initial_distribution,
+                                                transition_matrix,
+                                                log_likelihoods)
+
+    # Run the smoother backward in time
+    def _step(carry, args):
+        # Unpack the inputs
+        smoothed_probs_next = carry
+        t, filtered_probs = args
+
+        # Get parameters for time t
+        get = lambda x: x[t] if x.ndim == 3 else x
+        A = get(transition_matrix)
+
+        # Fold in the next state and renormalize
+        smoothed_probs = filtered_probs * (A @ smoothed_probs_next)
+        smoothed_probs /= smoothed_probs.sum()
+
+        # Compute p(z_t=i, z_{t+1}=j)
+        p_cross = filtered_probs[:, None] * A * smoothed_probs_next[None, :]
+        p_cross /= p_cross.sum()
+
+        return smoothed_probs, (smoothed_probs, p_cross)
+
+    # Run the HMM smoother
+    num_timesteps = len(log_likelihoods)
+    carry = filtered_probs[-1]
+    args = (jnp.arange(num_timesteps - 1, -1, -1), filtered_probs[:-1][::-1])
+    _, (rev_smoothed_probs, rev_smoothed_cross) = lax.scan(_step, carry, args)
+
+    # Reverse the arrays and return
+    smoothed_probs = jnp.row_stack([rev_smoothed_probs[::-1], filtered_probs[-1]])
+    smoothed_cross = rev_smoothed_cross[::-1]
+    return log_normalizer, smoothed_probs, smoothed_cross
+
+
+def hmm_posterior_mode(initial_distribution,
```

Comment:
Add comment that this is the Viterbi algorithm (so people can find this function if searching for it)